test_lots_of_deletes failing on *BSD

I'm getting segfaults on "test_lots_of_deletes" under FreeBSD:

```
Apr 23 13:24:18 [-] ----- test_lots_of_deletes -----

Apr 23 13:24:18 [-] Opened DB: 0x801407380.
Apr 23 13:24:20 [-] Records inserted: 1000000.
Apr 23 13:24:20 [-] Saw 516494 collisions.
./run_tests.sh: line 4: 98102 Segmentation fault      (core dumped) ./build/bin/oleg_test     test
gmake: *** [test] Error 139
```

OpenBSD seems to be more verbose:

```
Apr 23 15:29:47 [-] ----- test_lots_of_deletes -----

Apr 23 15:29:47 [-] Opened DB: 0xf48590ee800.
Apr 23 15:30:02 [-] Records inserted: 1000000.
Apr 23 15:30:02 [-] Saw 516494 collisions.
oleg_test(14280) in free(): error: chunk is already free 0xf4843086140
./run_tests.sh: line 4: 14280 Abort trap              (core dumped) ./build/bin/oleg_test         test
Makefile:89: recipe for target 'test' failed
gmake: *** [test] Error 134
```

--- a/src/oleg.c
+++ b/src/oleg.c
@@ -288,7 +288,7 @@ int ol_unjar_ds(ol_database *db, const char *key, size_t klen, unsigned char **d
             /* Decomperss with LZ4 if enabled */
             if (db->is_enabled(OL_F_LZ4, &db->feature_set)) {
                 int processed = 0;
-                processed = LZ4_decompress_fast((const char*)bucket->data_ptr,
+                processed = LZ4_decompress_fast((char *)bucket->data_ptr,
                                                 (char *)*data,
                                                 bucket->original_size);
                 check(processed == bucket->data_size, "Could not decompress data.");
@@ -316,18 +316,22 @@ static inline int _ol_reallocate_bucket(ol_database *db, ol_bucket *bucket,
         unsigned char *value, size_t vsize, const char *ct, const size_t ctsize) {
     debug("Reallocating bucket.");
 
+    unsigned char *data = NULL;
+
     /* Compress using LZ4 if enabled */
     size_t cmsize = 0;
     unsigned char* compressed = NULL;
     if (db->is_enabled(OL_F_LZ4, &db->feature_set)) {
-        compressed = malloc(vsize);
-        cmsize = (size_t)LZ4_compress((const char*)value, (char*)compressed,
+        int maxoutsize = LZ4_compressBound(vsize);
+        compressed = malloc(maxoutsize);
+        cmsize = (size_t)LZ4_compress((char*)value, (char*)compressed,
                                       (int)vsize);
-    }
+    } else {
+        data = realloc(bucket->data_ptr, vsize);
 
-    unsigned char *data = realloc(bucket->data_ptr, vsize);
-    if (memcpy(data, value, vsize) != data)
-        return 4;
+        if (memcpy(data, value, vsize) != data)
+            return 4;
+    }
 
     char *ct_real = realloc(bucket->content_type, ctsize+1);
     if (strncpy(ct_real, ct, ctsize) != ct_real)
@@ -415,15 +419,16 @@ int _ol_jar(ol_database *db, const char *key, size_t klen, unsigned char *value,
     new_bucket->original_size = vsize;
     if(db->is_enabled(OL_F_LZ4, &db->feature_set)) {
         /* Compress using LZ4 if enabled */
-        size_t cmsize = 0;
-        unsigned char *compressed = calloc(1, vsize);
-        cmsize = (size_t)LZ4_compress((const char*)value, (char*)compressed,
+        int maxoutsize = LZ4_compressBound(vsize);
+        unsigned char *compressed = calloc(1, maxoutsize);
+        size_t cmsize = (size_t)LZ4_compress((char*)value, (char*)compressed,
                                       (int)vsize);
-        unsigned char *ret = realloc(compressed, (size_t)cmsize);
+        check(cmsize > 0, "Compression failed");
+        unsigned char *ret = realloc(compressed, cmsize);
         check(ret != NULL, "Could not slim down memory for compressed data.");
 
         new_bucket->data_size = cmsize;
-        new_bucket->data_ptr = compressed;
+        new_bucket->data_ptr = ret;
     } else {
         new_bucket->data_size = vsize;
         unsigned char *data = calloc(1, vsize);

